report consistent position for all readers

yjh0502 opened the pull request:
It seems that in some cases, `StrRead`/`SliceRead` and `IoRead` report different error position for same input.
 - Add `from_reader` to `test_parse_err`/`test_parse_slice_err` to test an error position for `IoRead`
 - Fix an index tracking on `StrRead`/`SliceRead`
 - Modify `Deserializer` to track `fix_position` more accurate
 - Fix errors on testcases

yjh0502 commented on tests/test.rs:
```diff
@@ -1012,13 +1014,29 @@ fn test_parse_string() {
         (
             &[b'"', b'\n', b'"'],
-            "control character (\\u0000-\\u001F) found while parsing a string at line 1 column 1",
+            "control character (\\u0000-\\u001F) found while parsing a string at line 2 column 0",
```
I'm not sure `line 2 column 0` is a valid position or not. I left it with `line 2 column 0` because
 - `line 1 column 1` points to `"` which is wrong.
 - recovering a column information from `IoRead` seems quite expensive

dtolnay commented:
I am seeing a 2-3% performance regression from this. Is there any way to fix the problem without overhead?

[twitter.json](https://raw.githubusercontent.com/serde-rs/json-benchmark/9eae8499c19ecdf3531d72451e0172b56e570c99/data/twitter.json)

```rust
#[macro_use]
extern crate criterion;
extern crate serde;
extern crate serde_json;

use criterion::Criterion;

fn criterion_benchmark(c: &mut Criterion) {
    let j = std::fs::read_to_string("twitter.json").unwrap();
    c.bench_function("deserialize twitter.json", move |b| b.iter(|| {
        serde_json::from_str::<serde::de::IgnoredAny>(&j).unwrap();
    }));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
```

```console
deserialize twitter.json                        
                        time:   [1.0296 ms 1.0322 ms 1.0354 ms]
                        change: [+2.4354% +3.0627% +3.7191%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

![pdf](https://user-images.githubusercontent.com/1940490/46125516-4e93a700-c1de-11e8-8dee-c5a68f042126.png)

yjh0502 commented:
I updated the PR to maintain `peek` in reader, and here's a criterion result. Performance regression is still there, but seems smaller than first PR. I removed `peek_position` in updated PR and I'm not sure if it will break compatibility or not.

BTW I used `AMD Ryzen 7 2700X` for the benchmark.

```
deserialize twitter.json
                        time:   [1.1785 ms 1.1790 ms 1.1793 ms]
                        change: [+0.6121% +0.6896% +0.7694%] (p = 0.00 < 0.05)
                        Change within noise threshold.
```

![pdf](https://user-images.githubusercontent.com/554344/46128559-cf9a7f00-c26e-11e8-812c-fa0a6b2d19a1.png)

yjh0502 commented:
I get it. I'll close the PR for now and make a new PR for fixing some bugs on error location.
